packages/fx-core/src/core/middleware/projectMigrator.ts

```diff
@@ -0,0 +1,427 @@
+    return false;
+  }
+  const inputs = ctx.arguments[ctx.arguments.length - 1] as Inputs;
+  const fxExist = await fs.pathExists(path.join(inputs.projectPath as string, ".fx"));
+  if (!fxExist) {
+    return false;
+  }
+
+  const envFileExist = await checkFileExist(
+    path.join(inputs.projectPath as string, ".fx", "env.default.json")
+  );
+  const configDirExist = await fs.pathExists(
+    path.join(inputs.projectPath as string, ".fx", "configs")
+  );
+  const armParameterExist = await checkFileExist(
+    path.join(inputs.projectPath as string, ".fx", "configs", "azure.parameters.dev.json")
+  );
+  if (envFileExist && (!armParameterExist || !configDirExist)) {
+    return true;
+  }
+  return false;
+}
+
+function preCheckEnvEnabled() {
+  if (isMultiEnvEnabled() && isArmSupportEnabled() && isBicepEnvCheckerEnabled()) {
+    return true;
+  }
+  return false;
+}
+
+async function migrateArm(ctx: CoreHookContext) {
+  await removeBotConfig(ctx);
+  await generateArmTempaltesFiles(ctx);
+  await generateArmParameterJson(ctx);
+}
+
+async function removeBotConfig(ctx: CoreHookContext) {
+  const inputs = ctx.arguments[ctx.arguments.length - 1] as Inputs;
+  const fx = path.join(inputs.projectPath as string, `.${ConfigFolderName}`);
+  const envConfig = await fs.readJson(path.join(fx, "env.default.json"));
+  if (envConfig[ResourcePlugins.Bot]) {
+    delete envConfig[ResourcePlugins.Bot];
+    envConfig[ResourcePlugins.Bot] = { wayToRegisterBot: "create-new" };
+  }
+  await fs.writeFile(path.join(fx, "new.env.default.json"), JSON.stringify(envConfig, null, 4));
+}
+
+async function generateArmTempaltesFiles(ctx: CoreHookContext) {
+  const fakeCtx: CoreHookContext = { arguments: ctx.arguments };
+  const inputs = ctx.arguments[ctx.arguments.length - 1] as Inputs;
+  const core = ctx.self as FxCore;
+
+  const fx = path.join(inputs.projectPath as string, `.${ConfigFolderName}`);
+  const fxConfig = path.join(fx, InputConfigsFolderName);
+  const templateAzure = path.join(inputs.projectPath as string, "templates", "azure");
+  await fs.ensureDir(fx);
+  await fs.ensureDir(fxConfig);
+  await fs.ensureDir(templateAzure);
+  // load local settings.json
+  const loadRes = await loadProjectSettings(inputs);
+  if (loadRes.isErr()) {
+    throw ProjectSettingError();
+  }
+  const [projectSettings, projectIdMissing] = loadRes.value;
+  fakeCtx.projectSettings = projectSettings;
+  fakeCtx.projectIdMissing = projectIdMissing;
+
+  // load envinfo env.default.json
+  const targetEnvName = "default";
+  const result = await loadSolutionContext(
+    core.tools,
+    inputs,
+    fakeCtx.projectSettings,
+    fakeCtx.projectIdMissing,
+    targetEnvName,
+    inputs.ignoreEnvInfo
+  );
+  if (result.isErr()) {
+    throw SolutionConfigError();
+  }
+  fakeCtx.solutionContext = result.value;
+  // generate bicep files.
+  try {
+    await generateArmTemplate(fakeCtx.solutionContext);
+  } catch (error) {
+    return error;
```

Comment:
some code in this function throws an error(L332) and some code returns an error(this line), should we align the error handling?

Reply:
sure, this error need to throw, thanks

Reply:
We can throw error with confidence as it will be catched by `ErrorHandlerMW`